File: leetcode/889-construct-binary-tree-from-preorder-and-postorder-traversal/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func constructFromPrePost(pre []int, post []int) *TreeNode {
	if len(pre) == 0 || len(post) == 0 {
		return nil
	}
	node := &TreeNode{pre[0], nil, nil}
	if len(pre) == 1 || len(post) == 1 {
		return node
	}
	left_node_val := pre[1]
	left_pre := []int{}
	left_post := []int{}
	right_pre := []int{}
	right_post := []int{}
	for i := 0; i < len(post); i++ {
		if post[i] == left_node_val {
			left_pre = pre[1 : i+2]
			left_post = post[:i+1]
			if i+2 < len(pre) {
				right_pre = pre[i+2:]
				right_post = post[i+1 : len(post)-1]
			}
		}
	}
	node.Left = constructFromPrePost(left_pre, left_post)
	node.Right = constructFromPrePost(right_pre, right_post)
	return node
}

func main() {
}

File: leetcode/889-construct-binary-tree-from-preorder-and-postorder-traversal/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

"""
    1st: recursion

    pre     = [1,2,4,5,3,6,7]
    post    = [4,5,2,6,7,3,1]

    pre     = [1,2,4,5,3,6,7]
               ^ *     |
    post    = [4,5,2,6,7,3,1]
                   *     | ^

    * is the left subtree root
    we can get how many numbers in the left subtree by searching it in postorder
    then separate | the left subtree and right subtree


    Time    O(N^2)
    Space   O(N)
    64 ms, faster than 30.12%
"""


class Solution:
    def constructFromPrePost(self, pre: List[int], post: List[int]) -> TreeNode:
        if len(pre) == 0:
            return None
        val = pre[0]
        root = TreeNode(val)
        if len(pre) == 1:
            return root
        i = post.index(pre[1]) + 1
        root.left = self.constructFromPrePost(pre[1:i+1], post[:i])
        root.right = self.constructFromPrePost(pre[i+1:], post[i:-1])
        return root
